tests/unit_tests/sparse_blas/include/common_sparse_reference.hpp

```diff
@@ -0,0 +1,230 @@
+        const intType en = ia[i + 1] - indexing;
+        for (intType j = st; j < en; ++j) {
+            const intType col = ja[j] - indexing;
+            ia_t[col + 1]++;
+        }
+    }
+    // prefix sum to get official ia_t counts
+    ia_t[0] = indexing;
+    for (intType i = 0; i < a_ncols; ++i) {
+        ia_t[i + 1] += ia_t[i];
+    }
+
+    // second pass through data to fill transpose structure
+    for (intType i = 0; i < a_nrows; ++i) {
+        const intType st = ia[i] - indexing;
+        const intType en = ia[i + 1] - indexing;
+        for (intType j = st; j < en; ++j) {
+            const intType col = ja[j] - indexing;
+            const intType j_in_a_t = ia_t[col] - indexing;
+            ia_t[col]++;
+            ja_t[j_in_a_t] = i + indexing;
+            if (!structOnlyFlag) {
+                const fpType val = a[j];
+                a_t[j_in_a_t] = opVal(val, isConj);
+            }
+        }
+    }
+
+    // adjust ia_t back to original state after filling structure
+    for (intType i = a_ncols; i > 0; --i) {
+        ia_t[i] = ia_t[i - 1];
+    }
+    ia_t[0] = indexing;
+}
+
+// Transpose the given sparse matrix if needed
+template <typename fpType, typename intType>
+auto sparse_transpose_if_needed(const intType *ia, const intType *ja, const fpType *a,
+                                intType a_nrows, intType a_ncols, std::size_t nnz, intType indexing,
+                                oneapi::mkl::transpose transpose_val) {
+    std::vector<intType> iopa;
+    std::vector<intType> jopa;
+    std::vector<fpType> opa;
+    if (transpose_val == oneapi::mkl::transpose::nontrans) {
+        iopa.assign(ia, ia + a_nrows + 1);
+        jopa.assign(ja, ja + nnz);
+        opa.assign(a, a + nnz);
+    }
+    else if (transpose_val == oneapi::mkl::transpose::trans ||
+             transpose_val == oneapi::mkl::transpose::conjtrans) {
+        iopa.resize(static_cast<std::size_t>(a_ncols + 1));
+        jopa.resize(nnz);
+        opa.resize(nnz);
+        do_csr_transpose(transpose_val, iopa.data(), jopa.data(), opa.data(), a_nrows, a_ncols,
+                         indexing, ia, ja, a);
+    }
+    else {
+        throw std::runtime_error("unsupported transpose_val=" +
+                                 std::to_string(static_cast<char>(transpose_val)));
+    }
+    return std::make_tuple(iopa, jopa, opa);
+}
+
+template <typename fpType>
+auto dense_transpose_if_needed(const fpType *x, std::size_t outer_size, std::size_t inner_size,
+                               std::size_t ld, oneapi::mkl::transpose transpose_val) {
+    std::vector<fpType> opx;
+    if (transpose_val == oneapi::mkl::transpose::nontrans) {
+        opx.assign(x, x + outer_size * ld);
+    }
+    else {
+        opx.resize(outer_size * ld);
+        for (std::size_t i = 0; i < outer_size; ++i) {
+            for (std::size_t j = 0; j < inner_size; ++j) {
+                opx[i + j * ld] = x[i * ld + j];
+            }
+        }
```

Comment:
I suppose there's another option of keeping the size of the transposed matrix to have leading dimension `= inner_size`, unless we really want the transpose to also have a leading dimension `= ld`. Not saying we must make the following change, but is there a reason you've chosen to have `opx` keep the same leading dimension as `x`?
```suggestion
        opx.resize(outer_size * inner_size);
        for (std::size_t i = 0; i < outer_size; ++i) {
            for (std::size_t j = 0; j < inner_size; ++j) {
                opx[i + j * inner_size] = x[i * ld + j];
            }
        }
```

Reply:
I see what you mean. This would mean we would also need to remove the unused elements of `opx` for the `nontrans` case. It adds a bit more logic only to save a little bit of space on the host given the sizes that are tested. Either solution work the same for me so I have a small preference for not spending time on this.

Reply:
FYI I noticed some invalid write in this function when the matrix was transposed. I am finishing to debug some tests with the cuSPARSE backend and will update this PR when I can. I went with your suggestion of reducing the leading dimension to the minimum in this function.

Reply:
I rewrote part of the tests while working on the cuSPARSE backend. I found this avoided some invalided accesses in some cases and was easier to debug. https://github.com/oneapi-src/oneMKL/pull/500/commits/a79c5af025b99a77cd2c98164345ed24614e601a

Reply:
Resolving as this snippet has changed in the way you suggested.